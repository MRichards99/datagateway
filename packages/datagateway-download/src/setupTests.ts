import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

// below required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
function noOp(): void { }

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

export const flushPromises = (): Promise<void> => new Promise(setImmediate);

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: Function) => fn);
