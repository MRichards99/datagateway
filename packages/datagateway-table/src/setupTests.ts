import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Action } from 'redux';
import { StateType } from './state/app.types';
import { initialState } from './state/reducers/dgtable.reducer';
// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

// below required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
function noOp(): void {}

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

// these are used for testing async actions
export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getState = (): Partial<StateType> => ({ dgtable: initialState });
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: Function) => fn);
