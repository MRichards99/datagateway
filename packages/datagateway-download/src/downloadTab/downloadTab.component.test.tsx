import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import DownloadTabs from './downloadTab.component';
import { act } from 'react-dom/test-utils';
import { flushPromises } from '../setupTests';

describe('DownloadTab', () => {
  let shallow;
  let mount;

  beforeEach(() => {
    shallow = createShallow();
    mount = createMount();
  });

  afterEach(() => {
    mount.cleanUp();

    // Clear the session storage.
    sessionStorage.clear();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<DownloadTabs />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the previously used tab based on sessionStorage', async () => {
    let wrapper = mount(<DownloadTabs />);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(wrapper.exists('[aria-label="Download cart panel"]')).toBe(true);
    expect(
      wrapper.find('div[aria-label="Download cart panel"]').props().hidden
    ).toBe(false);

    // The tab index should be 0 for the cart tab.
    expect(sessionStorage.getItem('downloadStatusTab')).toEqual('0');

    await act(async () => {
      wrapper.find('button[aria-label="Downloads tab"]').simulate('click');

      await flushPromises();
      wrapper.update();

      expect(
        wrapper.find('div[aria-label="Download status panel"]').props().hidden
      ).toBe(false);
    });

    // The tab index should be 1 for the download tab.
    expect(sessionStorage.getItem('downloadStatusTab')).toEqual('1');

    // Recreate the wrapper and expect it to show the download tab.
    wrapper = mount(<DownloadTabs />);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(sessionStorage.getItem('downloadStatusTab')).toEqual('1');
    expect(wrapper.exists('[aria-label="Download status panel"]')).toBe(true);
    expect(
      wrapper.find('div[aria-label="Download status panel"]').props().hidden
    ).toBe(false);
  });

  it('shows the appropriate table when clicking between tabs', async () => {
    const wrapper = mount(<DownloadTabs />);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(wrapper.exists('[aria-label="Download cart panel"]')).toBe(true);
    expect(
      wrapper.find('div[aria-label="Download cart panel"]').props().hidden
    ).toBe(false);

    expect(wrapper.exists('[aria-label="Download status panel"]')).toBe(true);
    expect(
      wrapper.find('div[aria-label="Download status panel"]').props().hidden
    ).toBe(true);

    // Click on the downloads tab and the refresh downloads button.
    await act(async () => {
      wrapper.find('button[aria-label="Downloads tab"]').simulate('click');

      await flushPromises();
      wrapper.update();

      expect(
        wrapper.find('div[aria-label="Download status panel"]').props().hidden
      ).toBe(false);

      expect(
        wrapper.exists('[aria-label="Refresh download status table"]')
      ).toBe(true);

      wrapper
        .find('button[aria-label="Refresh download status table"]')
        .simulate('click');

      await flushPromises();
      wrapper.update();
    });

    // Return back to the cart tab.
    await act(async () => {
      wrapper.find('button[aria-label="Cart tab"]').simulate('click');
      await flushPromises();
      wrapper.update();

      expect(
        wrapper.find('div[aria-label="Download status panel"]').props().hidden
      ).toBe(true);
    });
  });
});
