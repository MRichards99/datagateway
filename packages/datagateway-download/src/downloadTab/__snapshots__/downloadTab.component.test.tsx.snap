// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadTab renders correctly 1`] = `
<WithStyles(ForwardRef(Paper))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Tabs))
    centered={true}
    indicatorColor="primary"
    onChange={[Function]}
    textColor="primary"
    value={0}
  >
    <WithStyles(ForwardRef(Tab))
      aria-controls="tabpanel-0"
      aria-label="Cart tab"
      id="tab-0"
      label="Cart"
    />
    <WithStyles(ForwardRef(Tab))
      aria-controls="tabpanel-1"
      aria-label="Downloads tab"
      id="tab-1"
      label="Downloads"
    />
  </WithStyles(ForwardRef(Tabs))>
  <TabPanel
    aria-label="Download cart panel"
    index={0}
    value={0}
  >
    <DownloadCartTable
      statusLink={[Function]}
    />
  </TabPanel>
  <TabPanel
    aria-label="Download status panel"
    index={1}
    value={0}
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      spacing={1}
    >
      <WithStyles(ForwardRef(Grid))
        aria-label="Last updated time"
        item={true}
        xs={12}
      >
        <WithStyles(ForwardRef(Typography))
          component="h3"
          style={
            Object {
              "alignItems": "center",
              "display": "inline-flex",
              "float": "right",
            }
          }
          variant="subtitle1"
        >
          <BlackTooltip
            enterDelay={500}
            title="Refresh Downloads"
          >
            <WithStyles(ForwardRef(IconButton))
              aria-label="Refresh download status table"
              color="secondary"
              onClick={[Function]}
            >
              <RefreshIcon />
            </WithStyles(ForwardRef(IconButton))>
          </BlackTooltip>
          <p
            style={
              Object {
                "paddingLeft": "10px ",
              }
            }
          >
            <b>
              Last checked: 
            </b>
             
          </p>
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={true}
      >
        <DownloadStatusTable
          refreshTable={false}
          setLastChecked={[Function]}
          setRefreshTable={[Function]}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </TabPanel>
</WithStyles(ForwardRef(Paper))>
`;
